pkg/agent/flowexporter/connections/conntrack_connections_perf_test.go

```diff
@@ -0,0 +1,180 @@
+// +build !race
+
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connections
+
+import (
+	"crypto/rand"
+	"flag"
+	"fmt"
+	"math/big"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/flowexporter"
+	connectionstest "antrea.io/antrea/pkg/agent/flowexporter/connections/testing"
+	"antrea.io/antrea/pkg/agent/flowexporter/flowrecords"
+	interfacestoretest "antrea.io/antrea/pkg/agent/interfacestore/testing"
+	"antrea.io/antrea/pkg/agent/openflow"
+	proxytest "antrea.io/antrea/pkg/agent/proxy/testing"
+	queriertest "antrea.io/antrea/pkg/querier/testing"
+	k8sproxy "antrea.io/antrea/third_party/proxy"
+)
+
+const (
+	testNumOfConns        = 100000
+	testNumOfNewConns     = 10000
+	testNumOfDeletedConns = 10000
+)
+
+var (
+	connStore      *ConntrackConnectionStore
+	conns          []*flowexporter.Connection
+	mockConnDumper *connectionstest.MockConnTrackDumper
+)
+
+/*
+Sample output:
+    go test -test.v -run=BenchmarkPoll -test.benchmem -bench=. -memprofile memprofile.out -cpuprofile profile.out
+	goos: linux
+	goarch: amd64
+	pkg: antrea.io/antrea/pkg/agent/flowexporter/connections
+	cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
+	BenchmarkPoll
+		conntrack_connections_perf_test.go:63:
+			Summary:
+			Number of initial connections: 100000
+			Number of new connections/poll: 10000
+			Number of deleted connections/poll: 10000
+	BenchmarkPoll-2   	       1	1406326459 ns/op	164436400 B/op	 3721614 allocs/op
+	PASS
+	ok  	antrea.io/antrea/pkg/agent/flowexporter/connections	1.581s
+*/
+
+func BenchmarkPoll(b *testing.B) {
+	disableLogToStderr()
+	setupConntrackConnStore(b)
+	for n := 0; n < b.N; n++ {
+		mockConnDumper.EXPECT().DumpFlows(uint16(openflow.CtZone)).Return(conns, testNumOfConns, nil)
+		connStore.Poll()
+		conns = generateUpdatedConns(conns)
+	}
+	b.Logf("\nSummary:\nNumber of initial connections: %d\nNumber of new connections/poll: %d\nNumber of deleted connections/poll: %d\n", testNumOfConns, testNumOfNewConns, testNumOfDeletedConns)
+}
```

Comment:
can you try this:
```go
disableLogToStderr()
setupConntrackConnStore(b)
b.ResetTimer()
for n := 0; n < b.N; n++ {
		mockConnDumper.EXPECT().DumpFlows(uint16(openflow.CtZone)).Return(conns, testNumOfConns, nil)
		connStore.Poll()
                 b.StopTimer()
                 conns = generateUpdatedConns(conns)
                 b.StartTimer()
}
```

I feel like the performance of `Poll()` is not that great according to the results, and I wonder if connection generation accounts for a lot of measurement overhead.

Also because of how you designed the benchmark, it looks like we need to run `Poll` multiple times to get good results. So I would recommend reducing the number of connections if needed, to ensure that we have multiple iterations. I'm assuming that resource usage (CPU / memory) increases linearly with the number of connections?

Reply:
Yes, it makes sense to not counting generation of new connections into runtime and run the tests multiple iterations. Tested with different number of connections, CPU and memory are increasing linearly with number of conns:
| # new conns   | # iterations |  ns/op | B/op   | allocs/op |
| ------------------------- | ------- | --------- | --- | --- |
|4000  |    256	|   4738835 ns/op	|  483896 B/op	|   27244 allocs/op| 
|6000  |    206	|   5989825 ns/op	|  608971 B/op	|   36037 allocs/op| 
|8000  |    157	|   7205885 ns/op	|  741305 B/op	|   45061 allocs/op| 
|10000 |   123	|   8710284 ns/op	|  882714 B/op	 |  54307 allocs/op| 
|12000  |   97	|  11834417 ns/op	|  1037488 B/op	|   63868 allocs/op|